### PR TITLE
Add transactions to mempool

### PR DESCRIPTION
**Pull request by omerfirmak**

### CODE REVIEW

File: rpc/handlers.go [hunk 1]
```diff
@@ -1169,6 +1181,31 @@ func (h *Handler) AddTransaction(ctx context.Context, tx BroadcastedTransaction)
...
+		DeclaredClass: userClass,
+	})
+	if err != nil {
+		return nil, ErrValidationFailure.CloneWithData(err.Error())
+	}
+
+	res := &AddTxResponse{TransactionHash: userTxn.Hash()}
+	if tx.Type == TxnDeployAccount {
+		res.ContractAddress = core.ContractAddress(&felt.Zero, tx.ClassHash, tx.ContractAddressSalt, *tx.ConstructorCallData)
+	} else if userClass != nil {
```
Comment (joshklop) [inline rpc/handlers.go:1200]: Is this equivalent to `tx.Type == TxnDeclare || tx.Type == TxnDeploy`?

File: rpc/handlers.go [hunk 2]
```diff
@@ -1169,6 +1181,31 @@ func (h *Handler) AddTransaction(ctx context.Context, tx BroadcastedTransaction)
...
+	userTxn, userClass, _, err := adaptBroadcastedTransaction(tx, h.bcReader.Network())
+	if err != nil {
+		return nil, ErrInternal.CloneWithData(err.Error())
+	}
+	err = h.memPool.Push(&mempool.BroadcastedTransaction{
+		Transaction:   userTxn,
+		DeclaredClass: userClass,
+	})
+	if err != nil {
+		return nil, ErrValidationFailure.CloneWithData(err.Error())
```
Comment (joshklop) [inline rpc/handlers.go:1194]: The error could be an internal database error. Maybe we should have `mempool.ErrInvalid` to know when to return `ErrInternal` and `ErrValidationFailure`.

File: rpc/handlers_test.go [hunk 3]
```diff
@@ -2800,7 +2797,7 @@ func TestAddTransaction(t *testing.T) {
 	}
 
 	for description, test := range tests {
-		t.Run(description, func(t *testing.T) {
+		t.Run(description+"/gateway", func(t *testing.T) {
```
Comment (joshklop) [inline rpc/handlers_test.go:2800]: What do you think about:

```go
t.Run(description, func(t *testing.T) {
        t.Run("gateway", func(t *testing.T) {
                // ...
        })
        t.Run("mempool", func(t *testing.T) {
                // ...
        })
})
```

**Review by joshklop — COMMENTED** [review]
We probably don't need #1596 anymore.

File: rpc/handlers.go [hunk 4]
```diff
@@ -1169,6 +1181,31 @@ func (h *Handler) AddTransaction(ctx context.Context, tx BroadcastedTransaction)
 	}, nil
 }
 
+func (h *Handler) addToMempool(tx *BroadcastedTransaction) (*AddTxResponse, *jsonrpc.Error) {
+	userTxn, userClass, _, err := adaptBroadcastedTransaction(tx, h.bcReader.Network())
+	if err != nil {
+		return nil, ErrInternal.CloneWithData(err.Error())
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply rpc/handlers.go:1187]: We will need to do much more granular error handling like in `makeJSONErrorFromGatewayError`, we can leave it for a separate PR for now.

File: rpc/handlers.go (hunk 2 above)
Comment (omerfirmak) [reply rpc/handlers.go:1194]: similar to https://github.com/NethermindEth/juno/pull/1622#discussion_r1439200956

File: rpc/handlers.go (hunk 1 above)
Comment (omerfirmak) [reply rpc/handlers.go:1200]: > tx.Type == TxnDeclare

Probably only this, `TxnDeploy` is not supported anymore.

**Comment by omerfirmak** [comment]
Merging assuming these were all the comment @joshklop had, we can revisit any part of this PR again as needed.

**Pull request merged**